### PR TITLE
Exit probs mapping

### PR DESCRIPTION
**Pull request by tsvisabo**
- This PR includes an implementation for tip selection that rather than performing random-walk whenever 
asking for a tip, this implementation maps all tips to their exit probabilities and allows sampling from them
- Allows multiple - relatively cheap tip selections until tangle is updated
- Allows querying for a tip's exit probability

### CODE REVIEW

File: utils/containers/hash/hash_map.c.tpl [hunk 1]
```diff
@@ -60,3 +60,11 @@ void hash_to_{TYPE}_map_free(hash_to_{TYPE}_map_t *const map) {
   }
   *map = NULL;
 }
+
+void hash_to_{TYPE}_map_keys(hash_to_{TYPE}_map_t *const map, hash243_set_t * const keys){
+  hash_to_{TYPE}_map_entry_t *curr_entry = NULL;
+  hash_to_{TYPE}_map_entry_t *tmp_entry = NULL;
+  HASH_ITER(hh, *map, curr_entry, tmp_entry) {
+  hash243_set_add(keys,curr_entry->hash);
```
Comment (th0br0) [inline utils/containers/hash/hash_map.c.tpl]: indent

File: consensus/exit_probability_randomizer/exit_prob_map.c [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+ * Private functions
+ */
+
+static void iota_consensus_exit_prob_map_add_exit_probs(
+    hash_to_double_map_t **const hash_to_probs, flex_trit_t const *const hash,
+    double value) {
+  hash_to_double_map_entry_t *curr_approver_entry = NULL;
+  if (hash_to_double_map_find(*hash_to_probs, hash, &curr_approver_entry)) {
+    curr_approver_entry->value += value;
+
```
Comment (th0br0) [inline consensus/exit_probability_randomizer/exit_prob_map.c]: remove

File: consensus/exit_probability_randomizer/exit_prob_map.c [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+retcode_t iota_consensus_exit_prob_map_randomize(
+    ep_randomizer_t const *const randomizer,
+    exit_prob_transaction_validator_t *const ep_validator,
+    cw_calc_result *const cw_result, flex_trit_t const *const ep,
+    flex_trit_t *tip) {
+  retcode_t ret;
+  ep_prob_map_randomizer_t *prob_randomizer =
+      (ep_prob_map_randomizer_t *const)randomizer;
```
Comment (th0br0) [inline consensus/exit_probability_randomizer/exit_prob_map.c]: remove const?

File: consensus/exit_probability_randomizer/exit_prob_map.c [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+  double rand_weight;
```
Comment (th0br0) [inline consensus/exit_probability_randomizer/exit_prob_map.c]: init to default value?

Comment (th0br0) [reply consensus/exit_probability_randomizer/exit_prob_map.c]: or move declaration to l101

File: consensus/exit_probability_randomizer/exit_prob_map.c [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+    hash_to_double_map_t *const hash_to_exit_probs,
+    hash_to_double_map_t *const hash_to_trans_probs) {
+  retcode_t ret;
+  bool ep_is_valid = false;
+  hash243_queue_t queue = NULL;
+  hash_to_indexed_hash_set_entry_t *aps = NULL;
+  hash_to_double_map_entry_t *curr_tx_entry = NULL;
+  hash243_set_entry_t *approver_entry = NULL;
+  hash243_set_entry_t *tmp_entry = NULL;
+  size_t num_approvers;
```
Comment (th0br0) [inline consensus/exit_probability_randomizer/exit_prob_map.c:129]: init to default values?

File: consensus/exit_probability_randomizer/exit_prob_map.h [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    hash_to_double_map_t *const hash_to_trans_probs);
+
+/**
+ * Virtual distructor
+ *
+ * @param exit_probability_randomizer The base class
+ *
+ * @return void
+ */
+
```
Comment (th0br0) [inline consensus/exit_probability_randomizer/exit_prob_map.h]: remvoe the whitelines between docs & method declaration?

File: consensus/exit_probability_randomizer/exit_prob_map.h [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+  hash_to_double_map_t transition_probs;
+};
+
+void iota_consensus_exit_prob_map_reset(
+    ep_randomizer_t *const exit_probability_randomizer);
+
+void iota_consensus_exit_prob_map_init(ep_randomizer_t *const randomizer);
+
+/**
+ * Randomize a tip selection based on exit probabilities
```
Comment (th0br0) [inline consensus/exit_probability_randomizer/exit_prob_map.h:43]: Could you also add who needs to own a lock if any to these comments?

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: consensus/exit_probability_randomizer/exit_prob_map.c (hunk 5 above)
Comment (tsvisabo) [reply consensus/exit_probability_randomizer/exit_prob_map.c:129]: `num_approvers` does not have a default value, and both vars won't be read before being set

File: consensus/exit_probability_randomizer/exit_prob_map.h (hunk 7 above)
Comment (tsvisabo) [reply consensus/exit_probability_randomizer/exit_prob_map.h:43]: Added doc, but it does make sense to add a lock to the class

**Review by th0br0 — APPROVED** [review]

**Pull request merged**